File: src/main/java/infinitespire/monsters/Nightmare.java
package infinitespire.monsters;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.evacipated.cardcrawl.modthespire.lib.SpireConfig;
import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.actions.GameActionManager;
import com.megacrit.cardcrawl.actions.animations.AnimateFastAttackAction;
import com.megacrit.cardcrawl.actions.common.*;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.cards.status.VoidCard;
import com.megacrit.cardcrawl.core.CardCrawlGame;
import com.megacrit.cardcrawl.core.Settings;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.helpers.CardLibrary;
import com.megacrit.cardcrawl.localization.MonsterStrings;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import com.megacrit.cardcrawl.powers.InvinciblePower;
import com.megacrit.cardcrawl.powers.MetallicizePower;
import com.megacrit.cardcrawl.powers.StrengthPower;
import com.megacrit.cardcrawl.vfx.cardManip.ExhaustCardEffect;
import com.megacrit.cardcrawl.vfx.combat.PowerBuffEffect;
import infinitespire.InfiniteSpire;
import infinitespire.effects.BlackCardEffect;
import infinitespire.powers.KeyMirrorPower;
import infinitespire.powers.RealityShiftPower;
import infinitespire.powers.SpireBlightPower;

import java.util.Random;

public class Nightmare extends AbstractMonster {

	public static final String ID = "Nightmare";
	public static final MonsterStrings monsterStrings = CardCrawlGame.languagePack.getMonsterStrings("Nightmare");
	public static final String NAME = monsterStrings.NAME;
	public static final String[] MOVES = Nightmare.monsterStrings.MOVES;
	public static final String[] DIALOG = Nightmare.monsterStrings.DIALOG;
	public static int timesDefeated = 0;
	public static int timesNotReceivedBlackCard = 0;

	private int textureIndex = 0;

	private float spriteEffect;
	private boolean firstTurn = true;
	private boolean hasActivated = false;
	private boolean isAlpha;
	private int attackDmg;
	private int slamDmg;
	private int debuffDmg;
	private int blockCount;
	private int effectCount;
	private boolean isStrong;
	private float portalRotation = 0.0f;

	public Nightmare() {
		this(false);
	}

	public Nightmare(boolean isAlpha) {
		super(NAME, ID, 200, -0.0F, -10.0F, 160f, 300f, null);
		this.isAlpha = isAlpha;
		this.type = EnemyType.ELITE;
		this.dialogX = -160.0f * Settings.scale;
		this.dialogY = 40f * Settings.scale;
		this.img = InfiniteSpire.getTexture("img/infinitespire/monsters/nightmare/nightmare-1.png");
		int hpAmount = 200;

		if(this.isAlpha){
			this.name = "Nightmare Alpha";
			hpAmount += 100;
		}

		if(CardCrawlGame.playerName.equals("fiiiiilth")){
			this.name = "Niiiiightmare";
		}

		this.debuffDmg = 10;

		if(AbstractDungeon.bossCount > 1 || timesDefeated > 0) {
			isStrong = true;
			hpAmount *= 1.5f;
			this.attackDmg = 5;
			this.slamDmg = 25;
			this.blockCount = 30;
			this.effectCount = 4; //effectively 5
		} else {
			this.attackDmg = 5;
			this.slamDmg = 15;
			this.blockCount = 20;
			this.effectCount = 2; //effectively 3
		}

		if(Settings.hasSapphireKey) {
			hpAmount += 50;
		}

		this.setHp(hpAmount);

		this.damage.add(new DamageInfo(this, this.attackDmg));
		this.damage.add(new DamageInfo(this, this.slamDmg));
		this.damage.add(new DamageInfo(this, this.debuffDmg));
	}

	public static void save(SpireConfig config) {
		config.setInt("nightmareTimesDefeated", timesDefeated);
		config.setInt("nightmareTimesNotReceivedBlackCard", timesNotReceivedBlackCard);
	}

	public static void load(SpireConfig config) {
		timesDefeated = config.getInt("nightmareTimesDefeated");
		timesNotReceivedBlackCard = config.getInt("nightmareTimesNotReceivedBlackCard");
	}

	public static void clear(){
		timesDefeated = 0;
		timesNotReceivedBlackCard = 0;
	}

	@Override
	public void die() {
		super.die();
		timesDefeated++;
		timesNotReceivedBlackCard++;
	}

	@Override
	public void render(SpriteBatch sb) {
		if(this.isAlpha) {
			sb.setColor(Color.WHITE.cpy());
			sb.draw(InfiniteSpire.getTexture(
				"img/infinitespire/screen/portal.png"),
				this.hb.cX - 128f, this.hb.cY - 128f, 128f, 128f,
				256f, 256f, 1.5f, 1.5f,
				portalRotation * 90f, 0, 0, 256, 256,
				false, false);
		}
		super.render(sb);
	}

	@Override
	public void update() {
		super.update();
		this.portalRotation += Gdx.graphics.getDeltaTime(); //look if you get a crash on this line you been sitting on that screen way too long and honestly i just cant be bothered to loop this float so it never crashes
		spriteEffect += Gdx.graphics.getDeltaTime();
		this.drawY += (Math.sin(spriteEffect) / (10f * Settings.scale));

		if(Math.round(spriteEffect * 10) % 3 == 0) {
			if((new Random()).nextInt(3) == 0) {
				this.textureIndex += 1;
				if(textureIndex > 2) {
					textureIndex = 0;
				}
				this.img = InfiniteSpire.getTexture("img/infinitespire/monsters/nightmare/nightmare-" + (textureIndex + 1) +".png");
			}
		}
	}

	@Override
	public void usePreBattleAction() {
		GameActionManager manager = AbstractDungeon.actionManager;

		if(Settings.hasRubyKey) {
			manager.addToBottom(new ApplyPowerAction(this, this, new KeyMirrorPower(this, KeyMirrorPower.KeyColor.RUBY)));
		}

		if(Settings.hasSapphireKey) {
			manager.addToBottom(new ApplyPowerAction(this, this, new KeyMirrorPower(this, KeyMirrorPower.KeyColor.SAPPHIRE)));
		}

		if(Settings.hasEmeraldKey) {
			manager.addToBottom(new ApplyPowerAction(this, this, new KeyMirrorPower(this, KeyMirrorPower.KeyColor.EMERALD)));
		}

		if(this.isAlpha){
			manager.addToBottom(new ApplyPowerAction(this, this, new RealityShiftPower(this), 50));

			if(Settings.hasEmeraldKey) {
				manager.addToBottom(new ApplyPowerAction(this, this, new InvinciblePower(this, this.maxHealth / 3)));
			}
		}

		if(Settings.hasEmeraldKey) {
			manager.addToBottom(new ApplyPowerAction(this, this, new StrengthPower(this, 2), 2));
		}

		if(isStrong) {

			if(timesDefeated > 0) {
				manager.addToBottom(new ApplyPowerAction(this, this,
					new MetallicizePower(this, 5 * timesDefeated)));
			}

			if(timesDefeated > 1) {
				manager.addToBottom(new ApplyPowerAction(this, this,
					new StrengthPower(this, timesDefeated - 1)));
			}
		}
	}

	@Override
	public void takeTurn() {
		switch(this.nextMove) {
			case 1:
				AbstractDungeon.actionManager.addToBottom(new ApplyPowerAction(AbstractDungeon.player, this, new SpireBlightPower(AbstractDungeon.player)));
				if(!this.hasPower("is_Reality_Shift")) {
					AbstractDungeon.actionManager.addToBottom(new ApplyPowerAction(this, this, new RealityShiftPower(this), 50));
				}
				break;
			case 2:
				AbstractDungeon.actionManager.addToBottom(new AnimateFastAttackAction(this));
				for(int i = 0; i < this.effectCount; i++) {
					AbstractDungeon.actionManager.addToBottom(new DamageAction(AbstractDungeon.player, this.damage.get(0), AbstractGameAction.AttackEffect.SLASH_HORIZONTAL, true));
				}
				break;
			case 3:
				AbstractDungeon.actionManager.addToBottom(new DamageAction(AbstractDungeon.player, this.damage.get(1), AbstractGameAction.AttackEffect.BLUNT_HEAVY, true));
				AbstractDungeon.actionManager.addToBottom(new ApplyPowerAction(this, this, new StrengthPower(this, 1), 1));
				break;
			case 4:
				AbstractDungeon.actionManager.addToBottom(new GainBlockAction(this, this, this.blockCount));
				break;
			case 5:
				AbstractDungeon.actionManager.addToBottom(new AnimateFastAttackAction(this));
				for(int i = 0; i < this.effectCount; i++) {
					AbstractDungeon.actionManager.addToBottom(new DamageAction(AbstractDungeon.player, this.damage.get(0), AbstractGameAction.AttackEffect.SLASH_HORIZONTAL, true));
				}
				AbstractDungeon.actionManager.addToBottom(new HealAction(this, this, 25));
				break;
			case 6:
				AbstractDungeon.actionManager.addToBottom(new DamageAction(AbstractDungeon.player, this.damage.get(2), AbstractGameAction.AttackEffect.SLASH_VERTICAL));
				AbstractCard voidCard = CardLibrary.getCopy(VoidCard.ID).makeCopy();

				AbstractDungeon.actionManager.addToBottom(new MakeTempCardInDrawPileAction(voidCard, 2, true, true));
				break;
		}
		AbstractDungeon.actionManager.addToBottom(new RollMoveAction(this));
	}

	@Override
	protected void getMove(int num) {
		if(this.isAlpha) {
			getMoveAlpha(num);
		} else {
			getMoveBeta(num);
		}
	}

	public void getMoveAlpha(int num) {
		this.hasActivated = false;
		if(firstTurn) {
			if(Settings.hasSapphireKey) {
				this.setMove(Nightmare.MOVES[1], (byte) 2, Intent.ATTACK, this.damage.get(0).base, this.effectCount, true);
			} else {
				this.setMove(Nightmare.MOVES[3], (byte) 4, Intent.DEFEND);
			}
			firstTurn = false;
			return;
		}
		if(num > 20) {
			if(num % 3 == 0){
				this.setMove(Nightmare.MOVES[1], (byte) 2, Intent.ATTACK, this.damage.get(0).base, this.effectCount, true);
			} else if(Settings.hasRubyKey && num % 3 == 1) {
				this.setMove((byte) 6, Intent.ATTACK_DEBUFF, this.damage.get(2).base);
			} else {
				this.setMove(Nightmare.MOVES[2], (byte) 3, Intent.ATTACK_BUFF, this.damage.get(1).base);
			}
		} else {
			this.setMove(Nightmare.MOVES[3], (byte) 4, Intent.DEFEND);
		}
	}

	public void getMoveBeta(int num) {
		this.hasActivated = false;
		if(firstTurn) {
			this.setMove(Nightmare.MOVES[0], (byte) 1, Intent.MAGIC);
			firstTurn = false;
			return;
		}
		if (num > 20) {
			if(Settings.hasRubyKey && num % 3 == 0){
				this.setMove(Nightmare.MOVES[1], (byte) 2, Intent.ATTACK, this.damage.get(0).base, this.effectCount, true);
			} else {
				this.setMove(Nightmare.MOVES[2], (byte) 3, Intent.ATTACK_BUFF, this.damage.get(1).base);
			}
		} else {
			this.setMove(Nightmare.MOVES[3], (byte) 4, Intent.DEFEND);
		}
	}

	public void onEnoughDamageTaken() {
		this.effectCount++;
		if(AbstractDungeon.overlayMenu.endTurnButton.enabled) {
			if(this.isAlpha){
				this.onEnoughDamageTakenAlpha();
			} else {
				this.onEnoughDamageTakenBeta();
			}
		}
	}

	public void onEnoughDamageTakenAlpha() {
		AbstractDungeon.effectList.add(new PowerBuffEffect(this.hb.cX - this.animX, this.hb.cY + this.hb.height / 2.0f, ""));

		AbstractDungeon.actionManager.addToTop(new GainBlockAction(this, this, this.blockCount));
		AbstractDungeon.effectList.add(new BlackCardEffect());

		this.setMove((byte) 5, Intent.ATTACK_BUFF, this.damage.get(0).base, this.effectCount, true);
		this.createIntent();

		AbstractDungeon.actionManager.cardQueue.clear();
		for (final AbstractCard c : AbstractDungeon.player.limbo.group) {
			AbstractDungeon.effectList.add(new ExhaustCardEffect(c));
		}
		AbstractDungeon.player.limbo.group.clear();
		AbstractDungeon.player.releaseCard();
		AbstractDungeon.overlayMenu.endTurnButton.disable(true);
		this.hasActivated = true;
	}

	public void onEnoughDamageTakenBeta() {
		AbstractDungeon.effectList.add(new PowerBuffEffect(this.hb.cX - this.animX, this.hb.cY + this.hb.height / 2.0f, ""));
		AbstractDungeon.actionManager.addToTop(new GainBlockAction(this, this, this.blockCount));
		AbstractDungeon.effectList.add(new BlackCardEffect());

		this.setMove(Nightmare.MOVES[1], (byte) 2, Intent.ATTACK, this.damage.get(0).base, this.effectCount, true);
		this.createIntent();

		AbstractDungeon.actionManager.cardQueue.clear();
		for (final AbstractCard c : AbstractDungeon.player.limbo.group) {
			AbstractDungeon.effectList.add(new ExhaustCardEffect(c));
		}
		AbstractDungeon.player.limbo.group.clear();
		AbstractDungeon.player.releaseCard();
		AbstractDungeon.overlayMenu.endTurnButton.disable(true);
		this.hasActivated = true;
	}
}

File: src/main/kotlin/com/blanktheevil/infinitespire/monsters/Nightmare.kt
package com.blanktheevil.infinitespire.monsters

import actlikeit.savefields.BehindTheScenesActNum
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.Vector2
import com.blanktheevil.infinitespire.extensions.*
import com.blanktheevil.infinitespire.interfaces.Savable
import com.blanktheevil.infinitespire.models.SaveData
import com.blanktheevil.infinitespire.monsters.utils.Move
import com.blanktheevil.infinitespire.monsters.utils.setMove
import com.blanktheevil.infinitespire.powers.RealityShiftPower
import com.blanktheevil.infinitespire.textures.Textures
import com.blanktheevil.infinitespire.utils.*
import com.blanktheevil.infinitespire.vfx.BlackCardVfx
import com.blanktheevil.infinitespire.vfx.particles.BlackCardParticle
import com.blanktheevil.infinitespire.vfx.particlesystems.BlackCardParticleSystem
import com.blanktheevil.infinitespire.vfx.utils.VFXManager
import com.megacrit.cardcrawl.actions.AbstractGameAction
import com.megacrit.cardcrawl.actions.animations.AnimateFastAttackAction
import com.megacrit.cardcrawl.actions.common.DamageAction
import com.megacrit.cardcrawl.actions.common.GainBlockAction
import com.megacrit.cardcrawl.actions.common.RollMoveAction
import com.megacrit.cardcrawl.blights.Spear
import com.megacrit.cardcrawl.cards.DamageInfo
import com.megacrit.cardcrawl.core.CardCrawlGame
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import com.megacrit.cardcrawl.powers.StrengthPower
import com.megacrit.cardcrawl.powers.WeakPower
import com.megacrit.cardcrawl.vfx.combat.PowerBuffEffect
import java.util.*
import kotlin.math.roundToInt

class Nightmare :
    Monster(
      NAME,
      ID,
      MAX_HP,
      HB_X,
      HB_Y,
      HB_W,
      HB_H,
      IMG_URL
    ), Savable {
  companion object {
    val ID = "Nightmare".makeID()
    private val STRINGS = languagePack.getMonsterStrings(ID)
    private val NAME = STRINGS.NAME
    private val MOVES = STRINGS.MOVES
    private const val MAX_HP = 550
    private const val MIN_HP = 520
    private const val ASC_HP = 50
    private const val BLOCK_ON_TRIGGER = 35
    private const val HB_X = 0f
    private const val HB_Y = -10f
    private const val HB_W = 160f
    private const val HB_H = 300f
    private const val IMG_URL = ""
    private var timesDefeated = 0
    private var timesNotReceivedBlackCard = 0
  }

  var attackAmount = 5
  private var attackDamage = 5
  private var slamDamage = 30
  private var debuffAmount = 3
  private var buffAmount = 1
  private var blockAmount = 35
  private var realityShiftAmount = 50
  private var firstTurn = true
  private var spriteTimer = 0f
  private var sprintIndex = 0
  private var particleSystem = BlackCardParticleSystem(
    8,
    {
      val point = VFXManager.generateRandomPointAlongEdgeOfCircle(
        hb.cX,
        hb.cY,
        200f.scale()
      )

      BlackCardParticle(
        point,
        1f,
        true,
        VFXManager.getVelocityToPoint(
          Vector2(hb.cX, hb.cY),
          point
        ).scl(0.75f.div(scale))
      )
    },
    { !isDying && !isDead }
  )
  private val realityShiftMove = Move(Intent.MAGIC, name = MOVES[0]) {
    it.applyPower(
      RealityShiftPower(it as Nightmare, 50)
    )
  }
  private val multiStrikeMove = Move(Intent.ATTACK, attackDamage, MOVES[1], attackAmount, true) {
    addToBot(
      AnimateFastAttackAction(it)
    )
    for (i in 0 until this.multiplier) {
      addToBot(DamageAction(
        player,
        it.damage[this.getByte().toInt()],
        AbstractGameAction.AttackEffect.SLASH_HORIZONTAL,
        true
      ))
    }
  }
  private val debuffBlockMove = Move(Intent.DEFEND_DEBUFF, name = MOVES[3]) {
    addToBot(GainBlockAction(it, blockAmount))
      for (i in 0 until debuffAmount) {
        player.applyPower(
          WeakPower(player, 1, true),
          source = it
        )
        player.applyPower(
          WeakPower(player, 1, true),
          source = it
        )
      }
  }
  private val slamMove = Move(Intent.ATTACK_BUFF, slamDamage, name = MOVES[2]) {
    addToBot(DamageAction(
          player,
          it.damage[this.getByte().toInt()],
          AbstractGameAction.AttackEffect.BLUNT_HEAVY
        ))
        applyPower(
          StrengthPower(it, buffAmount)
        )
  }

  init {
    type = EnemyType.BOSS

    if (AbstractDungeon.ascensionLevel > 7)
      setHp(MIN_HP.plus(ASC_HP), MAX_HP.plus(ASC_HP))
    else
      setHp(MIN_HP, MAX_HP)

    if (AbstractDungeon.ascensionLevel > 2) {
      multiStrikeMove.modify(damage = attackDamage + 1)
      slamMove.modify(damage = slamDamage + 5)
      this.debuffAmount += 1
    }

    if (CardCrawlGame.isInARun() && player.hasBlight(Spear.ID)) {
      realityShiftAmount *= 1 + player.getBlight(Spear.ID).counter
    }

    registerMove(realityShiftMove)
    registerMove(multiStrikeMove)
    registerMove(debuffBlockMove)
    registerMove(slamMove)

    this.img = Textures.monsters.get("nightmare/nightmare-1.png")
  }

  fun triggerRealityShiftAttack() {
    AbstractDungeon.effectList.add(
      PowerBuffEffect(
        this.hb.cX - this.animX,
        this.hb.cY + this.hb.height / 2f,
        "")
    )
    AbstractDungeon.effectList.add(
      BlackCardVfx()
    )
    addToTop(GainBlockAction(this, this, BLOCK_ON_TRIGGER))

    this.setMove(multiStrikeMove)
    this.createIntent()
    endPlayerTurn()
  }

  override fun damage(info: DamageInfo) {
    super.damage(info)
    if (info.owner != null && info.type != DamageInfo.DamageType.THORNS && info.output > 0) {
      particleSystem.addParticles(100) {
        val point = VFXManager.generateRandomPointAlongEdgeOfCircle(
          hb.cX,
          hb.cY,
          200f.scale()
        )

        BlackCardParticle(
          point,
          1f,
          true,
          VFXManager.getVelocityAwayFromPoint(
            Vector2(hb.cX, hb.cY),
            point
          ).scl(0.75f.div(scale))
        )
      }
    }
  }

  override fun update() {
    spriteTimer += deltaTime

    if (spriteTimer.times(10f).roundToInt().rem(3) == 0) {
      if ((Random()).nextInt(3) == 0) {
        if (++sprintIndex > 2) {
          sprintIndex = 0
        }
        img = Textures.monsters.get(
          "nightmare/nightmare-${sprintIndex + 1}.png"
        )
      }
    }

    particleSystem.update()

    super.update()
  }

  override fun render(sb: SpriteBatch) {
    particleSystem.render(sb)
    super.render(sb)
  }

  override fun dispose() = doNothing()

  override fun die() {
    var actNum = BehindTheScenesActNum.getActNum()
    BehindTheScenesActNum.setActNum(--actNum)
    log.info("Set Current Act Num: $actNum")
    super.die()
  }

  override fun getMove(roll: Int) {
    if (firstTurn) {
      this.setMove(realityShiftMove)
      firstTurn = false
      return
    }
    when {
      roll < 29 -> this.setMove(multiStrikeMove)
      roll > 69 -> this.setMove(debuffBlockMove)
      else -> this.setMove(slamMove)
    }
  }

  override fun beforeConfigSave(saveData: SaveData) {
    saveData.nightmareData.timesDefeated = timesDefeated
    saveData.nightmareData.timesNotReceivedBlackCard = timesNotReceivedBlackCard
  }

  override fun afterConfigLoad(saveData: SaveData) {
    timesDefeated = saveData.nightmareData.timesDefeated
    timesNotReceivedBlackCard = saveData.nightmareData.timesNotReceivedBlackCard
  }

  override fun clearData(saveData: SaveData) {
    timesDefeated = 0
    timesNotReceivedBlackCard = 0
  }
}